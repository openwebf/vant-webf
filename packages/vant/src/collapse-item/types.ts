import type { ComponentPublicInstance } from 'vue';
import type { CollapseItemProps } from './CollapseItem';

export type CollapseItemExpose = {
  toggle: (newValue?: boolean) => void;
};

export type CollapseItemInstance = ComponentPublicInstance<
  CollapseItemProps,
  CollapseItemExpose
>;
