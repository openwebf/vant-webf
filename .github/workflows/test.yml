name: CI

on:
  push:
    branches:
      - '**'

  pull_request:
    branches:
      - dev

  workflow_dispatch:

jobs:
  lint:
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v2

    - name: Install pnpm
      run: npm i pnpm@7 -g

    - uses: actions/setup-node@v2
      with:
        node-version: '16'
        cache: 'pnpm'

    - name: Install dependencies
      run: pnpm install

    - name: Run linter
      run: npm run lint

  test:
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v2

    - name: Install pnpm
      run: npm i pnpm@7 -g

    - uses: actions/setup-node@v2
      with:
        node-version: '16'
        cache: 'pnpm'

    - name: Install dependencies
      run: pnpm install

    - name: Run test cases
      run: npm test

    - name: Upload coverage to Codecov  
      uses: codecov/codecov-action@v1
      with:
        token: ${{ secrets.CODECOV_TOKEN }}

  build:
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v2

    - name: Install pnpm
      run: npm i pnpm@7 -g

    - uses: actions/setup-node@v2
      with:
        node-version: '16'
        cache: 'pnpm'

    - name: Install dependencies
      run: pnpm install

    - name: Build
      run: npm run build
