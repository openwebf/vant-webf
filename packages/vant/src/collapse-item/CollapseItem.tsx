import {
  ref,
  watch,
  computed,
  nextTick,
  defineComponent,
  type ExtractPropTypes,
} from 'vue';

// Utils
import { cellSharedProps } from '../cell/Cell';
import {
  pick,
  extend,
  truthProp,
  numericProp,
  createNamespace,
} from '../utils';
import { COLLAPSE_KEY } from '../collapse/Collapse';

// Composables
import { raf, doubleRaf, useParent } from '@vant/use';
import { useExpose } from '../composables/use-expose';
import { useLazyRender } from '../composables/use-lazy-render';

// Components
import { Cell } from '../cell';

const [name, bem] = createNamespace('collapse-item');

const CELL_SLOTS = ['icon', 'title', 'value', 'label', 'right-icon'] as const;

const collapseItemProps = extend({}, cellSharedProps, {
  name: numericProp,
  isLink: truthProp,
  disabled: Boolean,
  readonly: Boolean,
  lazyRender: truthProp,
});

export type CollapseItemProps = ExtractPropTypes<typeof collapseItemProps>;

export default defineComponent({
  name,

  props: collapseItemProps,

  setup(props, { slots }) {
    const wrapperRef = ref<HTMLElement>();
    const contentRef = ref<HTMLElement>();
    const { parent, index } = useParent(COLLAPSE_KEY);

    if (!parent) {
      if (process.env.NODE_ENV !== 'production') {
        console.error(
          '[Vant] <CollapseItem> must be a child component of <Collapse>.'
        );
      }
      return;
    }

    const name = computed(() => props.name ?? index.value);
    const expanded = computed(() => parent.isExpanded(name.value));

    const show = ref(expanded.value);
    const lazyRender = useLazyRender(() => show.value || !props.lazyRender);

    const onTransitionEnd = () => {
      if (!expanded.value) {
        show.value = false;
      } else if (wrapperRef.value) {
        wrapperRef.value.style.height = '';
      }
    };

    watch(expanded, (value, oldValue) => {
      if (oldValue === null) {
        return;
      }

      if (value) {
        show.value = true;
      }

      // Use raf: flick when opened in safari
      // Use nextTick: closing animation failed when set `user-select: none`
      const tick = value ? nextTick : raf;

      tick(() => {
        if (!contentRef.value || !wrapperRef.value) {
          return;
        }

        const { offsetHeight } = contentRef.value;
        if (offsetHeight) {
          const contentHeight = `${offsetHeight}px`;
          wrapperRef.value.style.height = value ? '0' : contentHeight;

          // use double raf to ensure animation can start
          doubleRaf(() => {
            if (wrapperRef.value) {
              wrapperRef.value.style.height = value ? contentHeight : '0';
            }
          });
        } else {
          onTransitionEnd();
        }
      });
    });

    const toggle = (newValue = !expanded.value) => {
      parent.toggle(name.value, newValue);
    };

    const onClickTitle = () => {
      if (!props.disabled && !props.readonly) {
        toggle();
      }
    };

    const renderTitle = () => {
      const { border, disabled, readonly } = props;
      const attrs = pick(
        props,
        Object.keys(cellSharedProps) as Array<keyof typeof cellSharedProps>
      );

      if (readonly) {
        attrs.isLink = false;
      }
      if (disabled || readonly) {
        attrs.clickable = false;
      }

      return (
        <Cell
          v-slots={pick(slots, CELL_SLOTS)}
          role="button"
          class={bem('title', {
            disabled,
            expanded: expanded.value,
            borderless: !border,
          })}
          aria-expanded={String(expanded.value)}
          onClick={onClickTitle}
          {...attrs}
        />
      );
    };

    const renderContent = lazyRender(() => (
      <div
        v-show={show.value}
        ref={wrapperRef}
        class={bem('wrapper')}
        onTransitionend={onTransitionEnd}
      >
        <div ref={contentRef} class={bem('content')}>
          {slots.default?.()}
        </div>
      </div>
    ));

    useExpose({ toggle, expanded, itemName: name });

    return () => (
      <div class={[bem({ border: index.value && props.border })]}>
        {renderTitle()}
        {renderContent()}
      </div>
    );
  },
});
