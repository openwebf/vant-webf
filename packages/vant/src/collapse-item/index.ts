import { withInstall } from '../utils';
import _CollapseItem from './CollapseItem';

export const CollapseItem = withInstall(_CollapseItem);
export default CollapseItem;
export type { CollapseItemProps } from './CollapseItem';
export type { CollapseItemInstance } from './types';

declare module 'vue' {
  export interface GlobalComponents {
    VanCollapseItem: typeof CollapseItem;
  }
}
