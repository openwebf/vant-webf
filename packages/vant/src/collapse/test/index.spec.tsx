import { defineComponent, ref } from 'vue';
import { Collapse } from '..';
import { CollapseItem } from '../../collapse-item';
import { later, mount } from '../../../test';

const Component = defineComponent({
  props: {
    accordion: Boolean,
    border: {
      type: Boolean,
      default: true,
    },
  },
  data() {
    return {
      active: this.accordion ? '' : [],
    };
  },
  render() {
    return (
      <Collapse
        v-model={this.active}
        border={this.border}
        accordion={this.accordion}
      >
        <CollapseItem title="a" name="first">
          content
        </CollapseItem>
        <CollapseItem title="b">content</CollapseItem>
        <CollapseItem title="c">content</CollapseItem>
      </Collapse>
    );
  },
});

test('should update active value when title is clicked', async () => {
  const wrapper = mount(Component);

  const titles = wrapper.findAll('.van-collapse-item__title');
  titles[0].trigger('click');
  expect(wrapper.vm.active).toEqual(['first']);

  await later();
  titles[1].trigger('click');
  expect(wrapper.vm.active).toEqual(['first', 1]);

  await later();
  titles[0].trigger('click');
  expect(wrapper.vm.active).toEqual([1]);

  wrapper.unmount();
});

test('should update active value when title is clicked in accordion mode', async () => {
  const wrapper = mount(Component, {
    props: {
      accordion: true,
    },
  });

  const titles = wrapper.findAll('.van-collapse-item__title');
  titles[0].trigger('click');
  expect(wrapper.vm.active).toEqual('first');

  titles[1].trigger('click');
  expect(wrapper.vm.active).toEqual(1);

  titles[0].trigger('click');
  expect(wrapper.vm.active).toEqual('first');

  await later();
  titles[0].trigger('click');
  expect(wrapper.vm.active).toEqual('');
});

test('should render slots of CollapseItem correctly', () => {
  const wrapper = mount({
    data() {
      return {
        active: [],
      };
    },
    render() {
      return (
        <Collapse v-model={this.active}>
          <CollapseItem
            v-slots={{
              icon: () => 'this is icon',
              title: () => 'this is title',
              value: () => 'this is value',
              label: () => 'this is label',
              'right-icon': () => 'this is right icon',
            }}
          />
        </Collapse>
      );
    },
  });

  expect(wrapper.html()).toMatchSnapshot();
});

test('should not render border when border prop is false', () => {
  const wrapper = mount(Component, {
    props: {
      border: false,
    },
  });

  expect(wrapper.find('.van-hairline--top-bottom').exists()).toBeFalsy();
});

test('should lazy render collapse content', async () => {
  const wrapper = mount({
    data() {
      return {
        active: [],
      };
    },
    render() {
      return (
        <Collapse v-model={this.active}>
          <CollapseItem title="a">content</CollapseItem>
          <CollapseItem title="b">
            <div class="foo" />
          </CollapseItem>
        </Collapse>
      );
    },
  });

  expect(wrapper.find('.foo').exists()).toBeFalsy();

  const titles = wrapper.findAll('.van-collapse-item__title');
  await titles[1].trigger('click');
  expect(wrapper.find('.foo').exists()).toBeTruthy();
});

test('should toggle collapse after calling the toggle method', async () => {
  const wrapper = mount({
    setup() {
      const itemA = ref();
      const itemB = ref();
      const active = ref([]);
      return {
        itemA,
        itemB,
        active,
      };
    },
    render() {
      return (
        <Collapse v-model={this.active}>
          <CollapseItem name="a" ref="itemA" />
          <CollapseItem name="b" ref="itemB" />
        </Collapse>
      );
    },
  });

  wrapper.vm.itemA.toggle();
  expect(wrapper.vm.active).toEqual(['a']);

  await later();
  wrapper.vm.itemB.toggle();
  expect(wrapper.vm.active).toEqual(['a', 'b']);

  wrapper.vm.itemB.toggle(false);
  expect(wrapper.vm.active).toEqual(['a']);

  wrapper.vm.itemA.toggle();
  expect(wrapper.vm.active).toEqual([]);
});

test('should toggle collapse after calling the toggle method in accordion mode', async () => {
  const wrapper = mount({
    setup() {
      const itemA = ref();
      const itemB = ref();
      const active = ref('');
      return {
        itemA,
        itemB,
        active,
      };
    },
    render() {
      return (
        <Collapse v-model={this.active} accordion>
          <CollapseItem name="a" ref="itemA" />
          <CollapseItem name="b" ref="itemB" />
        </Collapse>
      );
    },
  });

  wrapper.vm.itemA.toggle();
  await later();
  expect(wrapper.vm.active).toEqual('a');

  wrapper.vm.itemB.toggle();
  expect(wrapper.vm.active).toEqual('b');

  await later();
  wrapper.vm.itemB.toggle(false);
  expect(wrapper.vm.active).toEqual('');

  wrapper.vm.itemA.toggle();
  expect(wrapper.vm.active).toEqual('a');
});

test('should be readonly when using readonly prop', async () => {
  const wrapper = mount({
    data() {
      return {
        active: [],
      };
    },
    render() {
      return (
        <Collapse v-model={this.active}>
          <CollapseItem title="a" readonly>
            content
          </CollapseItem>
        </Collapse>
      );
    },
  });

  const titles = wrapper.findAll('.van-collapse-item__title');
  await titles[0].trigger('click');
  expect(wrapper.vm.active).toEqual([]);
  expect(wrapper.find('.van-collapse-item__title').classes()).not.toContain(
    'van-cell--clickable'
  );

  wrapper.unmount();
});
