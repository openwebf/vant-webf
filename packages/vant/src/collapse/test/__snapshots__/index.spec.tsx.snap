// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render slots of CollapseItem correctly 1`] = `
<div class="van-collapse van-hairline--top-bottom">
  <div class="van-collapse-item">
    <div class="van-cell van-cell--clickable van-collapse-item__title"
         role="button"
         tabindex="0"
         aria-expanded="false"
    >
      this is icon
      <div class="van-cell__title">
        this is title
        <div class="van-cell__label">
          this is label
        </div>
      </div>
      <div class="van-cell__value">
        this is value
      </div>
      this is right icon
    </div>
  </div>
</div>
`;
